### PR TITLE
Support (CREATE|DROP) INDEX CONCURRENTLY

### PR DESCRIPTION
**Pull request by jasonmp85**
Fixes: #1007

The changes here are as follow:

  * Modify `IndexStmt` and `DropStmt` to be deparsed on the master rather than a worker. Calling `worker_apply_shard_ddl_command` is not possible as it triggers a transaction and `CONCURRENTLY` statements must run outside a transaction
  * Remove checks preventing execution of `CONCURRENTLY` statement
  * Add flag to `DDLJob` to denote when transactions are prohibited
  * Add mode to force `Task`s to run serially rather than in parallel. `CREATE INDEX CONCURRENTLY` statements conflict with one another, _even on different tables_
  * Fix existing tests

# TODO

  - [x] Documentation pass
  - [x] Add back my failure testing
  - [x] Add slightly nicer error messaging
  - [x] (Possibly) mark index as invalid if any remote commands fail
  - [x] ~~Remove `IndextStmt` and `DropStmt` support from `RelayEventExtend`, etc.~~
  - [x] MX fix

### CODE REVIEW

**Comment by metdos** [comment]
> CREATE INDEX CONCURRENTLY statements conflict with one another, even on different tables

This is interesting. Do you know the reason?

File: src/backend/distributed/executor/multi_utility.c [hunk 1]
```diff
@@ -677,8 +679,9 @@ PlanIndexStmt(IndexStmt *createIndexStatement, const char *createIndexCommand)
 			{
 				DDLJob *ddlJob = palloc0(sizeof(DDLJob));
 				ddlJob->targetRelationId = relationId;
+				ddlJob->preventTransaction = createIndexStatement->concurrent;
 				ddlJob->commandString = createIndexCommand;
-				ddlJob->taskList = DDLTaskList(relationId, createIndexCommand);
+				ddlJob->taskList = IndexTaskList(relationId, createIndexStatement);
```
Comment (metdos) [inline src/backend/distributed/executor/multi_utility.c]: Would it be better to continue using worker_apply_shard_ddl_command() for regular index commands given that worker_apply_shard_ddl_command() is more native(?) and more tested?

```C
				if (createIndexStatement->concurrent)
				{
					ddlJob->taskList = IndexTaskList(relationId, createIndexStatement);
				} else {
					ddlJob->taskList = DDLTaskList(relationId, createIndexCommand);
				}
```

**Comment by metdos** [comment]
CTRL + C stops "create index concurrently" immediately. We see the master table with a successful index but some of the shards on workers don't get the index yet.

**Comment by metdos** [comment]
> (Possibly) mark index as invalid if any remote commands fail

This makes sense.

**Comment by metdos** [comment]
Currently, this fails on synchronized MX tables because we sync metadata in a transaction via SendCommandToWorkers().

```sql
postgres=# create index concurrently index_6 on lineitem (l_orderkey);
ERROR:  CREATE INDEX CONCURRENTLY cannot run inside a transaction block
CONTEXT:  while executing command on localhost:9700
```

One option is to send metadata commands for "create index concurrently" to workers without a transaction as you did for shards, however, the failure semantics will be complex.

I think it is ok to not support this for MX tables and to throw a more meaningful error message. However, we should check this with @ozgune first.

**Comment by jasonmp85** [comment]
Today I was mostly working on checking out the MX finding you posted. Replied in email to you, @ozgune, @sumedhpathak, and @marcocitus. Probably going to disable within MX for now.

You seemed to like the idea of marking the index as `invalid` if something goes wrong. That's still not perfect, since we can never have all the transactional guarantees we might like, but I think I'll add that back tomorrow.

Re `worker_apply_shard_ddl_command`, I actually think its support for `DropStmt` and `IndexStmt` should be removed. Here's my reasoning:

  * `CONCURRENTLY` doesn't prohibit any other flags, so to implement `CONCURRENTLY` using deparsing, I also had to support _all_ other flags
  * Because I support all flags, we'd have one path of code for non-`CONCURRENTLY`, and an entirely different path for `CONCURRENTLY` statements. This seems bad, since it doubles the bug surface for index features
  * I'm not even sold that `worker_apply_shard_ddl_command` is more "native", as I'd argue it's far more likely the parsenode representation changes than the SQL representation, which is pretty stable (and standardized)

So I think I'm leaning to switching to `deparse`-style for all index commands. The only open question is "what if an older coordinator node tries to call `worker_apply_shard_ddl_command` on a worker node running a newer Citus version?" IMO this is an unsupported configuration, so probably not something to worry about, but it was the only thing that occurred to me when I considered removing `IndexStmt` and `DropStmt` from `worker_apply_shard_ddl_command`.

File: src/backend/distributed/executor/multi_utility.c (hunk 1 above)
Comment (jasonmp85) [inline src/backend/distributed/executor/multi_utility.c]: Answered in a top-level comment.

**Comment by jasonmp85** [comment]
> This is interesting. Do you know the reason? 

It was _super_ hard to Google for this actually, but it turns out that `CONCURRENTLY` index creation takes place in two or three phases (see the code for more detail). Basically it creates the index, but in a state where it can't be used yet, commits, then starts building it, but doesn't allow it for certain operations yet, then commits, then flips on all operations. The docs say:

> When this option is used, PostgreSQL must perform two scans of the table, and in addition it must wait for all existing transactions that could potentially modify or use the index to terminate.

_all existing transactions that could potentially modify or use the index_. If you think about it for a sec, that actually means _all transactions_, since PostgreSQL has no clue what open transactions might do in their next command. So (I think?) it grabs `ShareLocks` on the `virtualxid` locks of all open transactions (!). This lock type isn't used for too much else (which is why Googling was hard), but transactions hold `ExclusiveLock`s on their own `virtualxid`. So a `CONCURRENTLY` creation will wait for all other transactions to complete, including other `CONCURRENTLY`s. So if I issue the commands in order, why do they get cycles?

> In a concurrent index build, the index is actually entered into the system catalogs in one transaction, then two table scans occur in two more transactions. Before each table scan, the index build must wait for existing transactions that have modified the table to terminate. After the second scan, the index build must wait for any transactions that have a snapshot (see Chapter 13) predating the second scan to terminate. Then finally the index can be marked ready for use, and the CREATE INDEX command terminates. Even then, however, the index may not be immediately usable for queries: in the worst case, it cannot be used as long as transactions exist that predate the start of the index build.

"the index build must wait for any transactions that have a snapshot predating the second scan". I think that's the phase where concurrent `CONCURRENTLY` deadlocks occur.

In short, I had been thinking of `CONCURRENTLY` as "faster" for some reason, but it's not. It has better operational characteristics (namely that `INSERT`s/`UPDATE`s, etc. can proceed), but can take much longer due to the multi-step lock acquisition. This will be compounded by the requirement that we issue the shard index creations serially.

**Comment by jasonmp85** [comment]
@metdos you can read a thread about `CREATE INDEX CONCURRENTLY` locking here: http://www.postgresql-archive.org/Simultaneous-index-creates-on-different-schemas-cause-deadlock-td5753045.html

It supposedly was relaxed by https://git.postgresql.org/gitweb/?p=postgresql.git;a=commit;h=c3d09b3bd23f5f65b5eb8124a3c7592dad85a50c but I'm not seeing that in practice.

**Comment by metdos** [comment]
> Today I was mostly working on checking out the MX finding you posted. Replied in email to you, @ozgune, @sumedhpathak, and @marcocitus. Probably going to disable within MX for now.

I think we should be fine to disable within MX for now. @ozgune, any objections on the product side?
> 
> You seemed to like the idea of marking the index as invalid if something goes wrong. That's still not perfect, since we can never have all the transactional guarantees we might like, but I think I'll add that back tomorrow.

My main concern is that, as a user, how will I know that my index is created completely on all shards? In the case of an error or "CTRL + C", PostgreSQL marks the index INVALID. Therefore, if we can't mark the index as INVALID, we should document this very well.

Maybe, we can also provide a udf to check the health of the index such as `verify_index_is_complete(index_name)`

> So I think I'm leaning to switching to deparse-style for all index commands.

Okay, your arguments make sense.

> The only open question is "what if an older coordinator node tries to call worker_apply_shard_ddl_command on a worker node running a newer Citus version?" IMO this is an unsupported configuration, so probably not something to worry about, but it was the only thing that occurred to me when I considered removing IndexStmt and DropStmt from worker_apply_shard_ddl_command.

Instead of removing them now, can we add depreciation comments and remove them later? This would be a problem if we start a rolling upgrade in the cluster from worker nodes.

**Comment by marcocitus** [comment]
> I think we should be fine to disable within MX for now. @ozgune, any objections on the product side?

By "disabling" we mean skipping the SendCommandToWorkers, right?

**Comment by metdos** [comment]
> By "disabling" we mean skipping the SendCommandToWorkers, right?

I was thinking erroring out, but skipping the SendCommandToWorkers and documenting the behavior for "create index concurrently" with MX is also fine.

**Comment by metdos** [comment]
> @metdos you can read a thread about CREATE INDEX CONCURRENTLY locking here: http://www.postgresql-archive.org/Simultaneous-index-creates-on-different-schemas-cause-deadlock-td5753045.html

Thanks for pointing out this. I like the comment of @anarazel noting that "... especially as nobody seems to have complained so far other than the OP ...". I think we will second the complain :)

> It supposedly was relaxed by https://git.postgresql.org/gitweb/?p=postgresql.git;a=commit;h=c3d09b3bd23f5f65b5eb8124a3c7592dad85a50c but I'm not seeing that in practice.

For PostgreSQL, I can easily replicate the deadlock on the same table (running index creation queries concurrently on the same table) but not on different ones (running index creation queries concurrently on different tables)  - though we can replicate the deadlock on different tables with your changes as I noted in my other comment.

For Citus, I couldn't replicate the deadlock on the master with distributed tables, either on the same table or on different tables.

**Comment by jasonmp85** [comment]
@metdos — Updated with many of the discussed changes. Couldn't outright remove `RelayEventExtendShardNames` or whatever because it's still called by the machinery that adds shards to existing tables.

Fixed MX, added MX tests (and noticed/fixed an MX test error), fixed up error messages, added partial failure tests, added `INVALID` marking, added function comments, etc.

File: src/backend/distributed/executor/multi_utility.c [hunk 2]
```diff
@@ -76,6 +77,7 @@
 #include "utils/palloc.h"
 #include "utils/rel.h"
 #include "utils/relcache.h"
+#include "utils/snapmgr.h"
```
Comment (metdos) [inline src/backend/distributed/executor/multi_utility.c]: Do we need to include snapmgr.h?

File: src/backend/distributed/transaction/worker_transaction.c [hunk 3]
```diff
@@ -69,6 +69,58 @@ SendCommandToWorkers(TargetWorkerSet targetWorkerSet, char *command)
...
+ */
+void
+SendBareCommandListToWorkers(TargetWorkerSet targetWorkerSet, List *commandList)
+{
+	List *workerNodeList = WorkerNodeList();
+	ListCell *workerNodeCell = NULL;
+	char *nodeUser = CitusExtensionOwnerName();
+	ListCell *commandCell = NULL;
+
+	if (XactModificationLevel > XACT_MODIFICATION_NONE)
```
Comment (metdos) [inline src/backend/distributed/transaction/worker_transaction.c]: Do we need this? PostgreSQL does not allow index concurrently in transactions anyway. Also, this error message is not visible because of PG_TRY()/PG_CATCH().

If we ditch this, we can also run `ExecuteSequentialTasksWithoutResults` before metadata sync above and reduce the window for the mx sync problem.

**Comment by metdos** [comment]
This locks COPY (expected given that we use transactions), but INSERT/UPDATE/DELETE are fine 👍

File: src/backend/distributed/transaction/worker_transaction.c [hunk 4]
```diff
@@ -69,6 +69,58 @@ SendCommandToWorkers(TargetWorkerSet targetWorkerSet, char *command)
...
 
 /*
+ * SendBareCommandListToWorkers sends a list of commands to a set of target
+ * workers in serial. Commands are committed immediately: new connections are
+ * always used and no transaction block is used (hence "bare"). The connections
+ * are made as the extension owner to ensure write access to the Citus metadata
+ * tables. Primarly useful for INDEX commands using CONCURRENTLY.
+ */
+void
+SendBareCommandListToWorkers(TargetWorkerSet targetWorkerSet, List *commandList)
```
Comment (jasonmp85) [inline src/backend/distributed/transaction/worker_transaction.c]: Elsewhere in the code we already use `COMMIT_PROTOCOL_BARE` to describe commands sent without any sort of transaction wrapping them, so it's out of analogy to that. I might leave this the way it is.

File: src/backend/distributed/transaction/worker_transaction.c (hunk 3 above)
Comment (jasonmp85) [inline src/backend/distributed/transaction/worker_transaction.c]: That's a reasonable point. I'll remove it. I'm not sure about calling this before `ExecuteSequentialTasks`, but I'll consider it.

File: src/include/distributed/multi_utility.h [hunk 5]
```diff
@@ -23,6 +23,7 @@ extern bool EnableDDLPropagation;
 typedef struct DDLJob
 {
 	Oid targetRelationId;      /* oid of the target distributed relation */
+	bool concurrent;           /* related to a CONCURRENTLY index command? */
```
Comment (jasonmp85) [inline src/include/distributed/multi_utility.h]: Hm. I'll change to `concurrentIndexCmd`, then.

File: src/backend/distributed/executor/multi_router_executor.c [hunk 6]
```diff
@@ -870,6 +870,29 @@ ExecuteModifyTasksWithoutResults(List *taskList)
 
 
 /*
+ * ExecuteSequentialTasksWithoutResults basically calls ExecuteModifyTasks in a
+ * loop in order to simulate sequential execution of a list of tasks. Useful in
+ * cases where issuing commands in parallel before waiting for results could
+ * result in deadlocks (such as CREATE INDEX CONCURRENTLY).
+ */
+int64
```
Comment (jasonmp85) [inline src/backend/distributed/executor/multi_router_executor.c]: That would make sense, yes. I copy-pasted from `ExecuteModifyTasksWithoutResults` and didn't realize I'd done that.

File: src/backend/distributed/executor/multi_router_executor.c [hunk 7]
```diff
@@ -870,6 +870,29 @@ ExecuteModifyTasksWithoutResults(List *taskList)
...
+ExecuteSequentialTasksWithoutResults(List *taskList)
```
Comment (jasonmp85) [inline src/backend/distributed/executor/multi_router_executor.c]: I'll change it to the latter.

File: src/backend/distributed/executor/multi_router_executor.c [hunk 8]
```diff
@@ -870,6 +870,29 @@ ExecuteModifyTasksWithoutResults(List *taskList)
...
+{
+	ListCell *taskCell = NULL;
+
+	foreach(taskCell, taskList)
+	{
+		Task *task = (Task *) lfirst(taskCell);
+		List *singleTask = list_make1(task);
+
+		ExecuteModifyTasks(singleTask, false, NULL, NULL);
```
Comment (jasonmp85) [inline src/backend/distributed/executor/multi_router_executor.c]: Sure.

File: src/backend/distributed/executor/multi_utility.c (hunk 2 above)
Comment (jasonmp85) [inline src/backend/distributed/executor/multi_utility.c]: I previously had push/pop snapshot calls in here. Removed, so I'll remove this `#include`.

File: src/backend/distributed/executor/multi_utility.c [hunk 9]
```diff
@@ -1987,15 +1984,49 @@ ExecuteDistributedDDLJob(DDLJob *ddlJob)
...
+	else
+	{
+		/* save old commit protocol to restore at xact end */
+		Assert(SavedMultiShardCommitProtocol == COMMIT_PROTOCOL_BARE);
+		SavedMultiShardCommitProtocol = MultiShardCommitProtocol;
+		MultiShardCommitProtocol = COMMIT_PROTOCOL_BARE;
+
+		PG_TRY();
+		{
+			if (shouldSyncMetadata)
```
Comment (jasonmp85) [inline src/backend/distributed/executor/multi_utility.c]: Sure, I'll swap them. Ultimately, the coordinator node will be marked `INVALID`, which close to as good as we can do, without a bunch of custom hackery.

File: src/backend/distributed/executor/multi_utility.c [hunk 10]
```diff
@@ -1987,15 +1984,49 @@ ExecuteDistributedDDLJob(DDLJob *ddlJob)
...
+
+			ExecuteSequentialTasksWithoutResults(ddlJob->taskList);
+		}
+		PG_CATCH();
+		{
+			ereport(ERROR,
+					(errmsg("CONCURRENTLY-enabled index command failed"),
+					 errdetail("CONCURRENTLY-enabled index commands can fail partially, "
+							   "leaving behind an INVALID index."),
+					 errhint("Use DROP INDEX IF EXISTS to remove the invalid index, then "
```
Comment (jasonmp85) [inline src/backend/distributed/executor/multi_utility.c]: Sure. Does dropping an index ever take that long? I'll suggest `CONCURRENTLY`.

File: src/backend/distributed/executor/multi_utility.c [hunk 11]
```diff
@@ -134,12 +136,15 @@ static bool IsAlterTableRenameStmt(RenameStmt *renameStatement);
 static void ExecuteDistributedDDLJob(DDLJob *ddlJob);
 static void ShowNoticeIfNotUsing2PC(void);
 static List * DDLTaskList(Oid relationId, const char *commandString);
+static List * IndexTaskList(Oid relationId, IndexStmt *indexStmt);
```
Comment (jasonmp85) [inline src/backend/distributed/executor/multi_utility.c]: Sure.

**Comment by jasonmp85** [comment]
Is this a :shipit: after these (minor?) things are fixed?

File: src/backend/distributed/executor/multi_utility.c (hunk 10 above)
Comment (metdos) [inline src/backend/distributed/executor/multi_utility.c]: Aha, I think it does take that long :)

File: src/backend/distributed/transaction/worker_transaction.c (hunk 3 above)
Comment (metdos) [inline src/backend/distributed/transaction/worker_transaction.c]: Just to be clear, I was referring to this comment for the order: https://github.com/citusdata/citus/pull/1287#discussion_r108928440

File: src/backend/distributed/transaction/worker_transaction.c (hunk 4 above)
Comment (metdos) [inline src/backend/distributed/transaction/worker_transaction.c]: Okay.

File: src/include/distributed/multi_utility.h (hunk 5 above)
Comment (metdos) [inline src/include/distributed/multi_utility.h]: Sounds good.

**Comment by metdos** [comment]
Hey @jasonmp85, I did a final pass, it looks good. After making changes you noted above, you can 
🚢  and 🚀  it!
